File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        int[] v = new int[1];
        v[0] = Integer.MIN_VALUE;
        help(root, v);
        return v[0];
    }

    public int help(TreeNode node, int[] v) {
        if (node == null) {
            return 0;
        }

        int left = help(node.left, v);
        int leftsum = Math.max(0, left);
        int right = help(node.right, v);
        int rightsum = Math.max(0, right);

        v[0] = Math.max(v[0], leftsum + rightsum + node.val);
        return node.val + Math.max(leftsum, rightsum);
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int mx = INT_MIN;
        int finalVal = help(root, mx);
        return mx;
    }
    int help(TreeNode* node, int& mx) {
        if (node == nullptr) {
            return 0;
        }

        int leftVal = max(0, help(node->left, mx));
        int rightVal = max(0, help(node->right, mx));

        mx = max(mx, node->val + leftVal + rightVal);

        return node->val + max(leftVal, rightVal);
    }
};